Add per-interpreter storage for `gil_safe_call_once_and_store`

XuehaiPan opened the pull request:
<!--
Title (above): please place [branch_name] at the beginning if you are targeting a branch other than master. *Do not target stable*.
It is recommended to use conventional commit format, see conventionalcommits.org, but not required.
-->
## Description

<!-- Include relevant issues or PRs here, describe what changed and why -->

Fixes #5926

For `!defined(PYBIND11_HAS_SUBINTERPRETER_SUPPORT)`, the previous behavior is unchanged.

For `defined(PYBIND11_HAS_SUBINTERPRETER_SUPPORT)`, the pybind11 `internals` is already a per-interpreter-dependent singleton stored in the interpreter's state dict. This PR adds a new map ~to the interpreter-dependent `internals`~ from the `gil_safe_call_once_and_store`'s address to the result, where the `gil_safe_call_once_and_store` is a static object that has a fixed address.

UPDATE: The storage map is moved to a separate `capsule` in the per-interpreter state dict instead of putting it as a member of `internals`. Because the per-interpreter `internals` are leaked in the program, the storage map is never GC-ed on (sub-)interpreter shutdown.

___

Note: CI logs show Python 3.12 emits a threading-related warning during subinterpreter tests:

```
2025-12-25T06:57:57.9327900Z [ RUN      ] GIL Subinterpreter
2025-12-25T06:57:57.9433000Z Exception ignored in: <module 'threading' from '/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/threading.py'>
2025-12-25T06:57:57.9433730Z Traceback (most recent call last):
2025-12-25T06:57:57.9440140Z   File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/threading.py", line 1602, in _shutdown
2025-12-25T06:57:57.9468980Z     assert tlock.locked()
2025-12-25T06:57:57.9469630Z            ^^^^^^^^^^^^^^
```

This appears during the "Single Subinterpreter" test and is 3.12-specific. Tests still pass. This seems to be a pre-existing issue with subinterpreter handling, not introduced by this PR.

## Suggested changelog entry:

<!-- Fill in the block below with the expected entry. Delete if no entry needed;
     but do not delete the header if an entry is needed! Will be collected via a script. -->

* Add per-interpreter storage for `gil_safe_call_once_and_store` to make it safe under multi-interpreters.


<!-- readthedocs-preview pybind11 start -->
----
📚 Documentation preview 📚: https://pybind11--5933.org.readthedocs.build/

<!-- readthedocs-preview pybind11 end -->

XuehaiPan commented on include/pybind11/detail/internals.h:
```diff
@@ -564,6 +596,15 @@ class internals_pp_manager {
+        // For example, a clean main interpreter that does not import any pybind11 module and then
+        // spawns multiple subinterpreters using `InterpreterPoolExecutor` that each imports a
+        // pybind11 module concurrently.
         if (get_num_interpreters_seen() > 1) {
```
Let me elaborate more about my failed test. I have a C++ singleton instance using `gil_safe_call_once_and_store` to ensure exactly one instance per-interpreter. With the patch in this PR, the call-once result is stored in the `internals` (`*get_pp()`), where the `internals` is stored in the interpreter's state dict.

```cpp
MyClass &get_singleton() {
    PYBIND11_CONSTINIT static py::gil_safe_call_once_and_store<MyClass> storage;
    return storage
        .call_once_and_store_result([]() -> MyClass {
            MyClass instance{};

            {
                // initialize
                ...
            }
            return instance;
        })
        .get_stored();
}
```

- **OK** - `test_import_in_subinterpreters_concurrently`: import the module in multiple subinterpreters concurrently, without any import for a pybind11 module in the main interpreter.
- **OK** - `test_import_in_subinterpreter_after_main`: import the module in the main interpreter, then import in a subinterpreter.
- **FAIL** - `test_import_in_subinterpreter_before_main`: import the module in a subinterpreter (the main interpreter does not import any pybind11 module yet), then import the module in the main interpreter.
    > The import succeeds in the subinterpreter, but the instance is ill-initialized in the main interpreter. If I remove the `get_num_interpreters_seen() > 1` entirely or initialize the atomic counter for seen interpreters with `2` instead of `0`, my test passes.

XuehaiPan commented on include/pybind11/detail/internals.h:
```diff
@@ -308,7 +337,12 @@ struct internals {
+            delete entry.second;
+        }
+        call_once_storage_map.clear();
+    }
```
> I suggest maybe just storing these items in capsules attached to the interpreter state dict, so that they can be destructed during the finalization/destruction of the owning interpreter.

The `internals` unique pointer is already stored as a `capsule` within the interpreter's state dict. Is there a functional difference between saving each attribute as an individual `capsule` versus simply letting the `internals` destructor handle its own attributes when the `capsule` is destroyed?

b-pass replied:
What's stored in the capsule is a pointer to the unique pointer, not the unique pointer itself.  The internals is destructed only after the interpreter has completed shutdown (see `pybind11::finalize_interpreter()` and `pybind11::subinterpreter::~subinterpreter()`).

So, in a case like this, where the interpreters are not created or destroyed by pybind11, the internals is never destructed.

XuehaiPan commented:
I have updated the approach to move the storage map as a separate `capsule` in the per-interpreter state dict instead of putting it in the `internals`. The PR description is also updated.

The CI tests are looking good so far. But I found some concurrency bugs for this patch in my downstream PR metaopt/optree#245. I'm not sure if it is coming from `pybind11` or the Python core itself.

- metaopt/optree#245

I will do further investigation about these.

------

Bug 1: Duplicate native enum registration.

```python
ImportError: pybind11::native_enum<...>("PyTreeKind") is already registered!
```

<details>
<summary>Test Case</summary>

```python
def test_import_in_subinterpreter_before_main():
    # OK
    check_script_in_subprocess(
        textwrap.dedent(
            """
            import contextlib
            import gc
            from concurrent import interpreters

            subinterpreter = None
            with contextlib.closing(interpreters.create()) as subinterpreter:
                subinterpreter.exec('import optree')

            import optree

            del optree, subinterpreter
            for _ in range(10):
                gc.collect()
            """,
        ).strip(),
        output='',
        rerun=NUM_FLAKY_RERUNS,
    )

    # FAIL
    check_script_in_subprocess(
        textwrap.dedent(
            f"""
            import contextlib
            import gc
            import random
            from concurrent import interpreters

            subinterpreter = subinterpreters = stack = None
            with contextlib.ExitStack() as stack:
                subinterpreters = [
                    stack.enter_context(contextlib.closing(interpreters.create()))
                    for _ in range({NUM_FUTURES})
                ]
                random.shuffle(subinterpreters)
                for subinterpreter in subinterpreters:
                    subinterpreter.exec('import optree')

            import optree

            del optree, subinterpreter, subinterpreters, stack
            for _ in range(10):
                gc.collect()
            """,
        ).strip(),
        output='',
        rerun=NUM_FLAKY_RERUNS,
    )

    # FAIL
    check_script_in_subprocess(
        textwrap.dedent(
            f"""
            import contextlib
            import gc
            import random
            from concurrent import interpreters

            subinterpreter = subinterpreters = stack = None
            with contextlib.ExitStack() as stack:
                subinterpreters = [
                    stack.enter_context(contextlib.closing(interpreters.create()))
                    for _ in range({NUM_FUTURES})
                ]
                random.shuffle(subinterpreters)
                for subinterpreter in subinterpreters:
                    subinterpreter.exec('import optree')

                import optree

            del optree, subinterpreter, subinterpreters, stack
            for _ in range(10):
                gc.collect()
            """,
        ).strip(),
        output='',
        rerun=NUM_FLAKY_RERUNS,
    )
```

</details>

<details>
<summary>Traceback</summary>

```pytb
__________________________________________________________________ test_import_in_subinterpreter_before_main ___________________________________________________________________

    def test_import_in_subinterpreter_before_main():
        check_script_in_subprocess(
            textwrap.dedent(
                """
                import contextlib
                import gc
                from concurrent import interpreters
    
                subinterpreter = None
                with contextlib.closing(interpreters.create()) as subinterpreter:
                    subinterpreter.exec('import optree')
    
                import optree
    
                del optree, subinterpreter
                for _ in range(10):
                    gc.collect()
                """,
            ).strip(),
            output='',
            rerun=NUM_FLAKY_RERUNS,
        )
    
>       check_script_in_subprocess(
            textwrap.dedent(
                f"""
                import contextlib
                import gc
                import random
                from concurrent import interpreters
    
                subinterpreter = subinterpreters = stack = None
                with contextlib.ExitStack() as stack:
                    subinterpreters = [
                        stack.enter_context(contextlib.closing(interpreters.create()))
                        for _ in range({NUM_FUTURES})
                    ]
                    random.shuffle(subinterpreters)
                    for subinterpreter in subinterpreters:
                        subinterpreter.exec('import optree')
    
                import optree
    
                del optree, subinterpreter, subinterpreters, stack
                for _ in range(10):
                    gc.collect()
                """,
            ).strip(),
            output='',
            rerun=NUM_FLAKY_RERUNS,
        )


concurrent/test_subinterpreters.py:267: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

script = "import contextlib\nimport gc\nimport random\nfrom concurrent import interpreters\n\nsubinterpreter = subinterpreters ...optree')\n\nimport optree\n\ndel optree, subinterpreter, subinterpreters, stack\nfor _ in range(10):\n    gc.collect()"

    def check_script_in_subprocess(script, /, *, output, env=None, cwd=TEST_ROOT, rerun=1):
        result = ''
        for _ in range(rerun):
            try:
                result = subprocess.check_output(
                    [sys.executable, '-Walways', '-Werror', '-c', script],
                    stderr=subprocess.STDOUT,
                    text=True,
                    encoding='utf-8',
                    cwd=cwd,
                    env={
                        key: value
                        for key, value in (env if env is not None else os.environ).items()
                        if not key.startswith(('PYTHON', 'PYTEST', 'COV_'))
                    },
                )
            except subprocess.CalledProcessError as ex:
>               raise CalledProcessError(ex.returncode, ex.cmd, ex.output, ex.stderr) from None
E               helpers.CalledProcessError: Command '['/home/PanXuehai/Projects/optree/venv/bin/python3', '-Walways', '-Werror', '-c', "import contextlib\nimport gc\nimport random\nfrom concurrent import interpreters\n\nsubinterpreter = subinterpreters = stack = None\nwith contextlib.ExitStack() as stack:\n    subinterpreters = [\n        stack.enter_context(contextlib.closing(interpreters.create()))\n        for _ in range(16)\n    ]\n    random.shuffle(subinterpreters)\n    for subinterpreter in subinterpreters:\n        subinterpreter.exec('import optree')\n\nimport optree\n\ndel optree, subinterpreter, subinterpreters, stack\nfor _ in range(10):\n    gc.collect()"]' returned non-zero exit status 1.
E               Output:
E               Traceback (most recent call last):
E                 File "<string>", line 16, in <module>
E                   import optree
E                 File "/home/PanXuehai/Projects/optree/optree/__init__.py", line 17, in <module>
E                   from optree import accessors, dataclasses, functools, integrations, pytree, treespec, typing
E                 File "/home/PanXuehai/Projects/optree/optree/accessors.py", line 25, in <module>
E                   import optree._C as _C
E               ImportError: pybind11::native_enum<...>("PyTreeKind") is already registered!

_          = 0
cwd        = PosixPath('/home/PanXuehai/Projects/optree/tests')
env        = None
output     = ''
rerun      = 8
result     = ''
script     = "import contextlib\nimport gc\nimport random\nfrom concurrent import interpreters\n\nsubinterpreter = subinterpreters ...optree')\n\nimport optree\n\ndel optree, subinterpreter, subinterpreters, stack\nfor _ in range(10):\n    gc.collect()"

helpers.py:187: CalledProcessError
```

</details>

------

Bug 2: Thread-safety issues for multiple-interpreters are ever created.

- Case 1: No multiple interpreters, only the main interpreter. Everything works fine.

    ```bash
    # Run tests separately
    make test PYTESTOPTS='-k "subinterpreter"'       # OK, no threading concurrency tests
    make test PYTESTOPTS='-k "not subinterpreter"'   # OK, the program only has the main interpreter
    ```

- Case 2: Create multiple interpreters, and destroy them before doing threading concurrency tests (only the main interpreter is alive during the concurrency tests).

    The C++ objects are locked with critical sections, and the GIL is enabled during the test.

    ~~~bash
    # There is only the main interpreter alive during threading concurrency tests
    #
    #     ```python
    #     with contextlib.closing(concurrent.interpreters.create()) as subintrepreter:
    #         # do something
    #         ...
    #
    #     # at this point, only the main interpreter is alive
    #
    #     # threading tests
    #     ...
    #     ```

    # Run tests alltogather
    make test  # subinterpreter tests all passed, but threading concurrency tests produce incorrect results
    ~~~

b-pass reviewed (changes requested):
I don't think these internals changes are a good idea.  Which things are being flaky and on which python versions?  I'd like to get to the bottom of those issues, but taking out big chunks of the internals management definitely makes me nervous!

b-pass commented:
I did some poking around in your new tests, and at least in the case of `MyEnum`, that name is used in another test so I think some of the errors you're getting are caused by that.  `MyClass` is also used elsewhere, but I think the `MyClass` errors might be a different problem with the tests.

It's possible there's a bug with the "import into sub-interpreter first" case ... we can probably detect that case during module init and increment the `get_num_interpreters_seen` to at least 2 if needed.  But I can't tell from everything going on in this PR whether that one is still a problem or not....

Are there tests in here not directly related to `gil_safe_call_once` that could be split out into a separate PR?

XuehaiPan commented:
> I did some poking around in your new tests, and at least in the case of `MyEnum`, that name is used in another test so I think some of the errors you're getting are caused by that. `MyClass` is also used elsewhere, but I think the `MyClass` errors might be a different problem with the tests.

The import test is in a completely new subprocess with a clean state. Modules built for other tests will not affect the correctness of the import tests for multiple-interpreters.

XuehaiPan commented:
> Are there tests in here not directly related to `gil_safe_call_once` that could be split out into a separate PR?

I can do that. But I want to point out that `PYBIND11_MODULE(..., py::multiple_interpreters::per_interpreter_gil())` is not usable at this moment due to duplicate registrations of C++ types.

I will comment out the newly added test in this PR and let this get merged first. Then open a new PR for concurrency issues for `internals`.

b-pass commented:
>  But I want to point out that PYBIND11_MODULE(..., py::multiple_interpreters::per_interpreter_gil()) is not usable at this moment due to duplicate registrations of C++ types.

The existing tests do define a class ("A") and import the module into multiple subinterpreters at the same time.

Do you have a test that shows this without using concurrent pools?  The concurrency makes it a lot more difficult to debug, and if it doesn't work at all it should be easy to demonstrate...?

b-pass commented on tests/test_with_catch/test_subinterpreter.cpp:
```diff
@@ -92,7 +117,10 @@ TEST_CASE("Single Subinterpreter") {
+// "Move Subinterpreter" test is disabled on free-threaded Python 3.14+ due to a hang
+// in Py_EndInterpreter() when the subinterpreter is destroyed from a different thread
+// than it was created on. See: https://github.com/pybind/pybind11/pull/5940
+#    if PY_VERSION_HEX >= 0x030D0000 && !(PY_VERSION_HEX >= 0x030E0000 && defined(Py_GIL_DISABLED))
```
The fix for this was merged to master, so this change should come out.

rwgk commented:
@XuehaiPan I think (strongly) we should split out the `tests/test_multiple_interpreters/` move as a separate, very conservative PR. I.e. the only theme should be the move, no tests added or modified. That'll make it easier to understand and debug the new code and tests.

Could you please let me know if you prefer doing that yourself, or me doing it? I believe with Cursor it'll be very straightforward. It'll probably do almost everything automatically, including merging the updated master into this PR (effectively a rebase, but I'd definitely do it with a regular merge commit, so that we have a coherent dev history here).

XuehaiPan commented:
> Could you please let me know if you prefer doing that yourself, or me doing it?

Done.

b-pass commented on include/pybind11/detail/internals.h:
For internals, this used to only run when the pointer already existed within the state dict.  It did not run for newly created internals.

rwgk replied:
Thanks for catching this! I spent a while carefully looking at the `atomic_get_or_create_in_state_dict` function, then created commit 1b6a857aa5d92f1239754b3bcc7832aa4e0c5994 to fix this issue, and apply some more minor changes. See commit message.

I still need to look at the latest versions of the new tests.

XuehaiPan commented:
@rwgk I think this one is ready for merge now. I will be working on the consistency issue in a follow-up PR:

- #5947

The diff will be updated once this PR gets merged.

rwgk commented:
@b-pass 

Is it expected that the attached code leaks?

My starting point was looking for leaks in the new
```
TEST_CASE("gil_safe_call_once_and_store per-interpreter isolation")
```

But after backtracking a few steps, it turns out that even the attached minimal loop leaks badly.

I'm testing on Ubuntu 24.04 with locally built Python 3.14.2, "default" and "freethreaded". The exact build commands I used are in a source code comment.

I'm looking with `top`. The `RES` size goes to several GB in a matter of a few seconds.

(This observation should not hold up this PR, it just makes it more tricky to sanity-check the new test.)

[while_true_scoped_subinterpreter.cpp](https://github.com/user-attachments/files/24335742/while_true_scoped_subinterpreter.cpp)

rwgk reviewed (approved):
I combed through with assistance from Cursor: LGTM

I'll wait for the CI to run to completion. If green, I'll merge, so we have a clean baseline for working on #5947
